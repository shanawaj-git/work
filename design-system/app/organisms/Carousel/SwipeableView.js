import React, { useState, useEffect } from 'react';
import MobileStepper from '@mui/material/MobileStepper';
import Button from 'atoms/Button';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import ArrowForwardIcon from '@mui/icons-material/ArrowForward';
import SwipeableViews from 'react-swipeable-views';
import { autoPlay } from 'react-swipeable-views-utils';
import PropTypes from 'prop-types';
import defaultContent from './defaultContent';

function SwipeableView(props) {
  const {
    direction,
    autoplay,
    position,
    nextButton,
    backButton,
    variant,
    content,
    className,
    maxSteps,
    nextIcon,
    backIcon,
    _ref,
    onIndexChanged,
  } = props;
  const isRtl = direction === 'rtl';
  const AutoPlaySwipeableViews = autoplay
    ? autoPlay(SwipeableViews)
    : SwipeableViews;

  const [activeStep, setActiveStep] = useState(0);

  useEffect(() => {
    if (onIndexChanged) {
      onIndexChanged(activeStep);
    }
  }, [activeStep]);

  const handleNext = () => {
    setActiveStep(prevActiveStep => prevActiveStep + 1);
  };

  const handleBack = () => {
    setActiveStep(prevActiveStep => prevActiveStep - 1);
  };

  const handleStepChange = step => {
    setActiveStep(step);
  };

  return (
    <div className={className} ref={_ref}>
      <AutoPlaySwipeableViews
        style={{
          flex: 1,
          flexDirection: 'flex',
          height: '100%',
        }}
        axis={isRtl ? 'x-reverse' : 'x'}
        index={activeStep}
        onChangeIndex={handleStepChange}
        enableMouseEvents
      >
        {content || defaultContent(activeStep)}
      </AutoPlaySwipeableViews>
      <MobileStepper
        variant={variant}
        steps={maxSteps}
        position={position}
        activeStep={activeStep}
        nextButton={
          nextButton ? (
            <Button
              onClick={handleNext}
              disabled={activeStep === maxSteps - 1}
              data-testid="next-btn"
            >
              {isRtl ? backIcon : nextIcon}
            </Button>
          ) : null
        }
        backButton={
          backButton ? (
            <Button
              onClick={handleBack}
              disabled={activeStep === 0}
              data-testid="back-btn"
            >
              {isRtl ? nextIcon : backIcon}
            </Button>
          ) : null
        }
      />
    </div>
  );
}

export default SwipeableView;
SwipeableView.defaultProps = {
  direction: 'rtl',
  autoplay: false,
  position: 'static',
  variant: 'dots',
  nextIcon: <ArrowForwardIcon />,
  backIcon: <ArrowBackIcon />,
};
SwipeableView.propTypes = {
  direction: PropTypes.string,
  position: PropTypes.string,
  variant: PropTypes.string,
  className: PropTypes.string,
  autoplay: PropTypes.bool,
  nextButton: PropTypes.bool,
  backButton: PropTypes.bool,
  nextIcon: PropTypes.element,
  backIcon: PropTypes.element,
  content: PropTypes.any,
  maxSteps: PropTypes.number,
  _ref: PropTypes.any,
  onIndexChanged: PropTypes.func,
};
