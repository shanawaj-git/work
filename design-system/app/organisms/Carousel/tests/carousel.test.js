import React from 'react';
import { render, fireEvent, wait } from 'react-testing-library';
import { act } from 'react-test-renderer';
import ArrowRightAltIcon from '@mui/icons-material/ArrowRightAlt';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';

import SwipeableView from '../index';
describe('Carousel', () => {
  it('Should import succesfully with default content', () => {
    const { container } = render(<SwipeableView />);
    expect(container).toBeTruthy();
  });
  it('renders correctly and match the snapshot with default variant', () => {
    const { container } = render(<SwipeableView />);
    expect(container).toMatchSnapshot();
  });
  it('renders correctly and match the snapshot with progress variant', () => {
    const { container } = render(<SwipeableView variant="progress" />);
    expect(container).toMatchSnapshot();
  });
  it('renders correctly and match the snapshot with custom next icon', () => {
    const { container } = render(
      <SwipeableView nextIcon={<ArrowRightAltIcon />} />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders correctly and match the snapshot with custom back icon', () => {
    const { container } = render(
      <SwipeableView backIcon={<ChevronLeftIcon />} />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders correctly with custom content', () => {
    act(() => {
      render(
        <SwipeableView
          variant="progress"
          content={[
            <div data-testid="slid01">slider-01</div>,
            <div>slider-02</div>,
            <div>slider-03</div>,
          ]}
        />,
      );
    });
    const firstSlider = document.querySelector('[data-testid="slid01"]');
    expect(firstSlider.textContent).toBe('slider-01');
  });
  it('renders correctly with next button', () => {
    act(() => {
      render(
        <SwipeableView
          variant="progress"
          content={[
            <div data-testid="slid01">slider-01</div>,
            <div data-testid="slid02">slider-02</div>,
            <div data-testid="slid03">slider-03</div>,
          ]}
          nextButton
        />,
      );
    });
    const nextbtn = document.querySelector('[type="button"]');
    expect(nextbtn).toBeTruthy();
  });
  it('renders correctly with back button', () => {
    act(() => {
      render(
        <SwipeableView
          variant="progress"
          content={[
            <div data-testid="slid01">slider-01</div>,
            <div data-testid="slid02">slider-02</div>,
            <div data-testid="slid03">slider-03</div>,
          ]}
          backButton
        />,
      );
    });
    const nextbtn = document.querySelector('[type="button"]');
    expect(nextbtn).toBeTruthy();
  });
  it('renders correctly with both buttons', () => {
    act(() => {
      render(<SwipeableView variant="progress" backButton nextButton />);
    });
    const btns = document.querySelectorAll('[type="button"]').length;
    expect(btns).toBe(2);
  });
  it('display slider one as a first layer', () => {
    act(() => {
      render(
        <SwipeableView
          variant="progress"
          nextButton
          content={[
            <div data-testid="slid01">slider-01</div>,
            <div data-testid="slid02">slider-02</div>,
            <div data-testid="slid03">slider-03</div>,
          ]}
          maxSteps={3}
        />,
      );
    });
    const activerSlider = document.querySelector('[aria-hidden="false"]');
    expect(activerSlider.textContent).toBe('slider-01');
  });
  it('display slider two after click on nextButton', () => {
    const tree = render(
      <SwipeableView
        variant="progress"
        nextButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
      />,
    );
    const btn = document.querySelector('[type="button"]');
    fireEvent.click(btn);
    tree.rerender(
      <SwipeableView
        variant="progress"
        nextButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
      />,
    );
    const activerSlider = document.querySelector('[aria-hidden="false"]');
    expect(activerSlider.textContent).toBe('slider-02');
  });
  it('display slider two after click on nextButton two times and backButton one time', () => {
    const tree = render(
      <SwipeableView
        variant="progress"
        nextButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
      />,
    );
    const nextBtn = document.querySelector('[type="button"]');
    fireEvent.click(nextBtn);
    fireEvent.click(nextBtn);

    tree.rerender(
      <SwipeableView
        variant="progress"
        backButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
      />,
    );
    const backBtn = document.querySelector('[type="button"]');
    fireEvent.click(backBtn);
    tree.rerender(
      <SwipeableView
        variant="progress"
        backButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
      />,
    );
    const activerSlider = document.querySelector('[aria-hidden="false"]');
    expect(activerSlider.textContent).toBe('slider-02');
  });

  it('triggers "onIndexChanged" with value "0" on first render', async () => {
    const onIndexChangedMockCallback = jest.fn();

    render(
      <SwipeableView
        variant="progress"
        nextButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
        onIndexChanged={onIndexChangedMockCallback}
      />,
    );

    await wait(onIndexChangedMockCallback);
    expect(onIndexChangedMockCallback.mock.calls[0][0]).toBe(0);
  });

  it('triggers "onIndexChanged" with value "1" on nextClick', () => {
    const onIndexChangedMockCallback = jest.fn();

    const tree = render(
      <SwipeableView
        variant="progress"
        nextButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
        onIndexChanged={onIndexChangedMockCallback}
      />,
    );
    const btn = tree.getByTestId('next-btn');
    fireEvent.click(btn);

    tree.rerender(
      <SwipeableView
        variant="progress"
        nextButton
        content={[
          <div data-testid="slid01">slider-01</div>,
          <div data-testid="slid02">slider-02</div>,
          <div data-testid="slid03">slider-03</div>,
        ]}
        maxSteps={3}
        onIndexChanged={onIndexChangedMockCallback}
      />,
    );

    expect(onIndexChangedMockCallback.mock.calls[1][0]).toBe(1);
  });
});
