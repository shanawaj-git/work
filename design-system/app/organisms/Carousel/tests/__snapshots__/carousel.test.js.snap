// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly and match the snapshot with custom back icon 1`] = `
<div>
  <div>
    <div
      style="overflow-x: hidden; flex: 1; flex-direction: flex; height: 100%;"
    >
      <div
        class="react-swipeable-view-container"
        style="flex-direction: row-reverse; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
      >
        <div
          aria-hidden="false"
          data-swipeable="true"
          style="width: 100%; flex-shrink: 0; overflow: auto;"
        >
          <div>
            <img
              alt="San Francisco – Oakland Bay Bridge, United States"
              class="MuiBox-root css-ijk13v"
              src="https://images.unsplash.com/photo-1537944434965-cf4679d1a598?auto=format&fit=crop&w=400&h=250&q=60"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiMobileStepper-root MuiMobileStepper-positionStatic css-rh92k-MuiPaper-root-MuiMobileStepper-root"
    >
      <div
        class="MuiMobileStepper-dots css-1384nzf-MuiMobileStepper-dots"
      >
        <div
          class="MuiMobileStepper-dot MuiMobileStepper-dotActive css-26w9jf-MuiMobileStepper-dot"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel renders correctly and match the snapshot with custom next icon 1`] = `
<div>
  <div>
    <div
      style="overflow-x: hidden; flex: 1; flex-direction: flex; height: 100%;"
    >
      <div
        class="react-swipeable-view-container"
        style="flex-direction: row-reverse; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
      >
        <div
          aria-hidden="false"
          data-swipeable="true"
          style="width: 100%; flex-shrink: 0; overflow: auto;"
        >
          <div>
            <img
              alt="San Francisco – Oakland Bay Bridge, United States"
              class="MuiBox-root css-ijk13v"
              src="https://images.unsplash.com/photo-1537944434965-cf4679d1a598?auto=format&fit=crop&w=400&h=250&q=60"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiMobileStepper-root MuiMobileStepper-positionStatic css-rh92k-MuiPaper-root-MuiMobileStepper-root"
    >
      <div
        class="MuiMobileStepper-dots css-1384nzf-MuiMobileStepper-dots"
      >
        <div
          class="MuiMobileStepper-dot MuiMobileStepper-dotActive css-26w9jf-MuiMobileStepper-dot"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel renders correctly and match the snapshot with default variant 1`] = `
<div>
  <div>
    <div
      style="overflow-x: hidden; flex: 1; flex-direction: flex; height: 100%;"
    >
      <div
        class="react-swipeable-view-container"
        style="flex-direction: row-reverse; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
      >
        <div
          aria-hidden="false"
          data-swipeable="true"
          style="width: 100%; flex-shrink: 0; overflow: auto;"
        >
          <div>
            <img
              alt="San Francisco – Oakland Bay Bridge, United States"
              class="MuiBox-root css-ijk13v"
              src="https://images.unsplash.com/photo-1537944434965-cf4679d1a598?auto=format&fit=crop&w=400&h=250&q=60"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiMobileStepper-root MuiMobileStepper-positionStatic css-rh92k-MuiPaper-root-MuiMobileStepper-root"
    >
      <div
        class="MuiMobileStepper-dots css-1384nzf-MuiMobileStepper-dots"
      >
        <div
          class="MuiMobileStepper-dot MuiMobileStepper-dotActive css-26w9jf-MuiMobileStepper-dot"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel renders correctly and match the snapshot with progress variant 1`] = `
<div>
  <div>
    <div
      style="overflow-x: hidden; flex: 1; flex-direction: flex; height: 100%;"
    >
      <div
        class="react-swipeable-view-container"
        style="flex-direction: row-reverse; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
      >
        <div
          aria-hidden="false"
          data-swipeable="true"
          style="width: 100%; flex-shrink: 0; overflow: auto;"
        >
          <div>
            <img
              alt="San Francisco – Oakland Bay Bridge, United States"
              class="MuiBox-root css-ijk13v"
              src="https://images.unsplash.com/photo-1537944434965-cf4679d1a598?auto=format&fit=crop&w=400&h=250&q=60"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiMobileStepper-root MuiMobileStepper-positionStatic css-rh92k-MuiPaper-root-MuiMobileStepper-root"
    >
      <span
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="NaN"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate MuiMobileStepper-progress css-1be5mm1-MuiLinearProgress-root-MuiMobileStepper-progress"
        role="progressbar"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(NaN%);"
        />
      </span>
    </div>
  </div>
</div>
`;
